test/support/utils_test_base.h

```diff
@@ -0,0 +1,695 @@
+
+    // Update data
+    virtual void update_data(UDTKind kind, TestValueType* __it_from, TestValueType* __it_to) override;
+};
+#endif // TEST_DPCPP_BACKEND_PRESENT
+
+#if TEST_DPCPP_BACKEND_PRESENT
+////////////////////////////////////////////////////////////////////////////////
+/// struct test_base_data_buffer - test source data for SYCL buffer
+template <typename TestValueType>
+struct test_base_data_buffer : test_base_data<TestValueType>
+{
+    struct Data
+    {
+        using TSourceData = sycl::buffer<TestValueType, 1>;
+
+        TSourceData   src_data_buf;     // SYCL buffer
+        ::std::size_t offset = 0;       // Offset in SYCL buffer
+
+        template<typename _Size>
+        Data(_Size __sz, ::std::size_t __offset)
+            : src_data_buf(sycl::range<1>(__sz + __offset))
+            , offset(__offset)
+        {
+        }
+    };
+    ::std::vector<Data> data;   // Vector of source test data:
+                                //  - 1 item for test1buffer;
+                                //  - 2 items for test2buffers;
+                                //  - 3 items for test3buffers
+
+    test_base_data_buffer(InitParams init);
+
+    sycl::buffer<TestValueType, 1>& get_buffer(UDTKind kind);
+
+    auto get_start_from(UDTKind kind)
+        -> decltype(oneapi::dpl::begin(data.at(enum_val_to_index(kind)).src_data_buf));
+
+// test_base_data
+
+    // Check that host buffering is required
+    virtual bool host_buffering_required() const override;
+
+    // Get test data
+    virtual TestValueType* get_data(UDTKind kind) override;
+
+    // Retrieve data
+    virtual void retrieve_data(UDTKind kind, TestValueType* __it_from, TestValueType* __it_to) override;
+
+    // Update data
+    virtual void update_data(UDTKind kind, TestValueType* __it_from, TestValueType* __it_to) override;
+};
+#endif // TEST_DPCPP_BACKEND_PRESENT
+
+////////////////////////////////////////////////////////////////////////////////
+/// struct test_base_data_sequence -  test source data for sequence (based on std::vector)
+template <typename TestValueType>
+struct test_base_data_sequence : test_base_data<TestValueType>
+{
+    struct Data
+    {
+        using TSourceData = Sequence<TestValueType>;
+
+        TSourceData   src_data_seq;     // Sequence
+        ::std::size_t offset = 0;       // Offset in sequence
+
+        Data(::std::size_t size, ::std::size_t __offset)
+            : src_data_seq(size)
+            , offset(__offset)
+        {
+        }
+    };
+    ::std::vector<Data> data;   // Vector of source test data:
+                                //  - 3 items for test_algo_three_sequences
+
+    test_base_data_sequence(InitParams init);
+
+    auto get_start_from(UDTKind kind)
+        -> decltype(data.at(enum_val_to_index(kind)).src_data_seq.begin());
+
+// test_base_data
+
+    // Check that host buffering is required
+    virtual bool host_buffering_required() const override;
```

Comment:
//TODO: consider to remove that virtual function as a redundant.  The polymorphic behavior should be defined within  'retrieve_data' and 'update_data' methods.

Reply:
Done